flang/include/flang/Optimizer/Dialect/FIRType.h

```diff
@@ -313,6 +313,9 @@ bool isBoxNone(mlir::Type ty);
 /// e.g. !fir.box<!fir.type<derived>>
 bool isBoxedRecordType(mlir::Type ty);
 
+/// Return true iff `ty` is a type that contains descriptor information.
+bool isTypeWithDescriptor(mlir::Type ty);
```

Comment:
Given this function might not see much use outside of the code added in this PR. Perhaps, we should have it as a local function/lambda instead?

If you feel like this is useful to have standalone, then we can keep it.

Reply:
It's used in a few different files in this case and I believe I was asked to make a utility function for this check earlier in this review, so I'd prefer to keep it as a utility function for the time being! Likely quite helpful if what classifies a descriptor type in Flang ever changes as well and it hopefully helps to make it clear what a descriptor type is represented as for newer implementors.